Add and refactor diffmerge tools

RussKie opened the pull request:
<!-- Please read CONTRIBUTING.md before submitting a pull request -->



## Proposed changes

- Refactor how diff/merge tools are defined and configured, each tool is moved into own type/class that defines all tool's
properties, such as name, executable, diff/merge commald line args, etc.
This will simlify how new tools are added, and allow easy deprecation of old tools. New versions of tools can easily be added too as new types.

- Streamline diff/merge tool related logic in how tools are configured by users in Configuration dialogs. All special handling for selected tools are removed (e.g. KDiff3 and TMerge). By default Git Extensions will have no awareness of diff/merge tools.
🎉  As a bonus - a user will be able to see what parameters each tool takes, and be able to configure custom scripts/executables to be executed as part of diff/merge invocation.

- Remove support for TortoiseGit prior 1.8 (TortoiseMerge.exe)

- Remove support for determining KDiff3 location from the registry.

- Remove "Keep backup (.orig) after merge" setting

## Back-compat

- Any pre-configured tools should just work. E.g. "semanticdiff" tool has been renamed to "semanticmerge" for consistency. However if it was configured in .gitconfig, it will continue to be correctly loaded in Git Extensions:
    ```
    [diff]
        tool = semanticdiff
        guitool = semanticdiff
    [difftool "semanticdiff"]
        path = C:/Users/russkie/AppData/Local/semanticmerge/semanticmergetool.exe
        cmd = \"C:/Users/russkie/AppData/Local/semanticmerge/semanticmergetool.exe\" -s \"$LOCAL\" -d \"$REMOTE\"
    [difftool]
        prompt = false
    [merge]
        tool = semanticmerge
    [mergetool "semanticmerge"]
        path = C:/Users/russkie/AppData/Local/semanticmerge/semanticmergetool.exe
        cmd = \"C:/Users/russkie/AppData/Local/semanticmerge/semanticmergetool.exe\" -s \"$REMOTE\" -d \"$LOCAL\" -b \"$BASE\" -r \"$MERGED\"
        trustExitCode = true
    [mergetool]
        prompt = false
        keepBackup = false
    ```

  ![image](https://user-images.githubusercontent.com/4403806/69494267-bbebb680-0f0d-11ea-9783-969d25bf259e.png)

## Screenshots:

- Open Git Extensions with no tool configured:
![image](https://user-images.githubusercontent.com/4403806/69494483-5b11ad80-0f10-11ea-882c-713d284c8848.png)

- Diff/merge tool configuration screen:
![image](https://user-images.githubusercontent.com/4403806/68767046-25e59f80-0674-11ea-9083-c81cd050b80e.png)

- Attempt to resolve a merge conflict without a diff/merge tool configured:
![image](https://user-images.githubusercontent.com/4403806/69494489-7e3c5d00-0f10-11ea-9f79-14936432e282.png)





## Test methodology <!-- How did you ensure quality? -->

- manual
- unit tests




----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

RussKie commented:
This is something I've started back in mid 2017, and it was lying around annoying me.
May as well finish it up....

@gerhardol this may help/align with #7009

gerhardol commented:
I would prefer #7009 first, as that aligns the behavior with the GE implementation, then rebase on master. I have considered ripping out som of the GE specifics and rely on Git handling only though, then most of this is a little overkill.

gerhardol commented:
I would like to go in the direction and remove most settings for merge/diff tools, to rely on Git instead. 
difftool.path may still be needed, but difftool.cmd will not

See for instance output of 
```git difftool --gui --tool-help```
A few need to be added like vscode (and TortoiseDiff?)
Tools the user have added should be seen too

The tricky is how to verify that a tool is setup (a test repo is one way).

This PR is not in conflict with my view, but parts of it will be reverted.

RussKie commented:
Thank you for pointing that out.
I have tried resetting .gitconfig and configuring git with `git difftool
--tool=<tool>` and it didn't appear to make any difference... Like at all.

I don't think our UI needs to change, and the logic is now significantly
simpler and cleaner than it used to be.

One thing I'm really puzzled about why we are using `diff.guitool` instead
of `diff.tool`, or why we are not using `merge.guitool`.

gerhardol commented:
merge.guitool is not used because it was added in Git 2.20 option and it was hard to add before the hardcoding for tools was removed.

git.difftool is a frontend to git.diff, the --gui is supposed to differ command line from terminal compatible from what I see, but in most cases there is no specific with guitool. I have used this personally to have a separate difftool from the commandline.

As I see the different "classes" of tools:
 * Git supported and found (blacklist vimdiff*) and "Git user-defined". path, cmd is not required, no editing (but save as new "tool")
 * GE built-in (not already listed by Git). Path required (but suggested by GE)
 * Git built in, not in path. Require only path
 * Custom tools, supply a default cmd - how to verify that it is working, a repo is needed? 

Checks for cmd, path has to be adapted

RussKie commented:
I don't see a problem (am I missing something?):

- GE knows about several tools, e.g. their exe file names, default installation location and commands. 
Just because git knows that too doesn't mean we need to remove our configs. 
We may supply more information that git may need, but is this a problem?
- Once a tool selected, GE tries to find an exe for the selected tool. If GE can't find it, likely git won't find it either.
The user will need to supply the path, or select a different tool.
- If the user configured a different tool that GE doesn't know about, GE will just show it. (this needs to be verified)

gerhardol commented:
> I don't see a problem (am I missing something?):

The problem is that difftool.<tool>.cmd and .path may not be needed and it is a little difficult to test the setup outside a repo.
The checks for .path, .cmd will have to be modified.

The rest (custom paths etc) is just details.

RussKie commented:
Should be good to go.

mstv commented on UnitTests/GitUITests/DiffMergeTools/DiffMergeToolsProviderTests.cs:
```diff
@@ -0,0 +1,164 @@
+﻿////using System;
```
What's your plan for this file?

RussKie replied:
Ah thanks! It was very late, and I had this file marked as "viewed", so I missed it.
It needs to go, I wrote it as an initial test harness.

gerhardol commented:
Will make formal review, I have looked at it and considered how to change to support "Git tools". 
The changes are mostly not in the changed code

gerhardol commented on GitCommands/DiffMergeTools/WinMerge.cs:
```diff
@@ -0,0 +1,25 @@
+        public override string ExeFileName => "winmergeu.exe";
+
+        /// <inheritdoc />
+        public override string MergeCommand => "-e -u -dl \"Original\" -dr \"Modified\" \"$MERGED\" \"$REMOTE\"";
```
case "winmerge":
                    return "\"" + exeFile + "\" -e -u  -wl -wr -fm -dl \"Mine: $LOCAL\" -dm \"Merged: $BASE\" -dr \"Theirs: $REMOTE\" \"$LOCAL\" \"$BASE\" \"$REMOTE\" -o \"$MERGED\"";

RussKie replied:
For a different version we need to add another `DiffMergeTool` definition

gerhardol replied:
This was a change in the defaults

RussKie replied:
Thank you

gerhardol commented on GitCommands/DiffMergeTools/TortoiseGitMerge.cs:
```diff
@@ -0,0 +1,23 @@
+    internal class TortoiseGitMerge : DiffMergeTool
+    {
+        /// <inheritdoc />
+        public override string ExeFileName => "TortoiseGitMerge.exe";
```
This removes support for tgit < 1.8 as well as using tsvn binary (unless that is changed too)
(just noting, probably OK)

RussKie replied:
If different versions of diff/merge tool have different exe names/different parameter sets they must be presented as different `DiffMergeTool` definitions.
This will allow tool-specific configurations and deprecation of out-dated tools in a clean way

gerhardol replied:
Should be documented in commit message

RussKie replied:
Added to the commit message

gerhardol commented on GitCommands/DiffMergeTools/SemanticMerge.cs:
```diff
@@ -0,0 +1,36 @@
+            string folder = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
+            return new[]
+            {
+                Path.Combine(folder, @"semanticmerge"),
```
Adding, OK

gerhardol commented on GitCommands/DiffMergeTools/SemanticMerge.cs:
```diff
@@ -0,0 +1,36 @@
+        public override string MergeCommand => "-s \"$REMOTE\" -d \"$LOCAL\" -b \"$BASE\" -r \"$MERGED\"";
+
+        /// <inheritdoc />
+        public override string Name => "semanticmerge";
```
semanticdiff for the difftool currently

RussKie replied:
In the latest version there is a single executable, naming them differently is not a good idea.

gerhardol replied:
Then this should be documented in commit message and reslease notes

RussKie replied:
This will continue to work as is - configured tools will be recognised even though they may not be in our official list:
![image](https://user-images.githubusercontent.com/4403806/69494267-bbebb680-0f0d-11ea-9783-969d25bf259e.png)

gerhardol replied:
This should be documented in the commit:
The diff tool names _semanticdiff_ and _tortoisemerge_ is now named _semanticmerge_ and _tmerge_ as the mergetools already was called.

gerhardol commented on GitCommands/DiffMergeTools/Kdiff3.cs:
```diff
@@ -0,0 +1,20 @@
+        /// <inheritdoc />
+        public override IEnumerable<string> SearchPaths => new[]
+        {
+            // regkdiff3path
```
Registry does not seem to be queried any more - maybe OK

RussKie replied:
Ah yes, I missed this...
Do we really want t bring it back?

gerhardol replied:
It is a functionality change, can probably be documented. KDiff3 is probably not the tool most install nowadays (it was a good addition when it was added for sure).

RussKie replied:
Noted the removal in the commit message

gerhardol commented on GitUI/CommandsDialogs/SettingsDialog/Pages/ChecklistSettingsPage.cs:
```diff
@@ -502,81 +482,42 @@ private bool CheckGitCmdValid()
-
-                    RenderSettingSet(MergeTool, MergeTool_Fix, string.Format(_customMergeToolXConfigured.Text, mergetool));
-                    return true;
-                }
```
This special handling is now removed. Probably OK

RussKie replied:
I don't want any more special handling, it makes the code incoherent mess.

gerhardol replied:
Neither do I, but the behavior changes, this at least must be documented in the commit message. It is very hard to investigate changes otherwise.

RussKie replied:
Done

gerhardol replied:
pushed?

RussKie commented on GitCommands/DiffMergeTools/DiffMerge.cs:
```diff
@@ -0,0 +1,23 @@
+        public override string ExeFileName => "DiffMerge.exe";
+
+        /// <inheritdoc />
+        public override string MergeCommand => "/m /r =\"$MERGED\" \"$LOCAL\" \"$BASE\" \"$REMOTE\"";
```
Good spot, I'll investigate

gerhardol replied:
I would prefer -merge and -result; the long names for readability and '-' to keep multi platform compatibilty in general (DiffMerge runs at least on Mac).

RussKie replied:
Done

RussKie commented:
Good points, I'll action them.
Thank you

On Tue, 19 Nov 2019 at 04:07, Gerhard Olsson <notifications@github.com>
wrote:

> *@gerhardol* commented on this pull request.
> ------------------------------
>
> In GitCommands/DiffMergeTools/WinMerge.cs
> <https://github.com/gitextensions/gitextensions/pull/7044#discussion_r347501224>
> :
>
> > @@ -0,0 +1,25 @@
> +using System.Collections.Generic;
> +
> +namespace GitCommands.DiffMergeTools
> +{
> +    internal class WinMerge : DiffMergeTool
> +    {
> +        /// <inheritdoc />
> +        public override string DiffCommand => "-e -u \"$LOCAL\" \"$REMOTE\"";
> +
> +        /// <inheritdoc />
> +        public override string ExeFileName => "winmergeu.exe";
> +
> +        /// <inheritdoc />
> +        public override string MergeCommand => "-e -u -dl \"Original\" -dr \"Modified\" \"$MERGED\" \"$REMOTE\"";
>
> This was a change in the defaults
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/7044?email_source=notifications&email_token=ABBTEXUZVM4WWWVDRSLJTG3QULDVNA5CNFSM4IPDDOBKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCL54NJI#discussion_r347501224>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXWX6BVBTRSJNV52MMDQULDVNANCNFSM4IPDDOBA>
> .
>

RussKie commented on GitCommands/DiffMergeTools/TortoiseGitMerge.cs:
```diff
@@ -0,0 +1,23 @@
+        public override string MergeCommand => "-base:\"$BASE\" -mine:\"$LOCAL\" -theirs:\"$REMOTE\" -merged:\"$MERGED\"";
+
+        /// <inheritdoc />
+        public override string Name => "tortoisemerge";
```
fixed, thank you

gerhardol replied:
Note: "tortoisemerge" used in FormResolveConflicts, both should be added (see also suggestion below)

RussKie commented:
I believe I addressed the review points. I have updated the commit messages and the PR description. Added few more screenshots.

Ready for another round.
Thank you

gerhardol commented:
Next steps:
 * Remove hardcoded use of .cmd and .path in FormResolveConflicts, the parameters are not required. A few manipulations for instance for Use2WayMerge() will have to be dropped. Or only provide those options when .cmd is available.
 * Make .cmd editable in settings
 * (possibly) add custom entries
 * (possibly) add Git built-in tools (automatically)

gerhardol commented on GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.cs:
```diff
@@ -133,118 +137,87 @@ protected override void PageToSettings()
+        private string BrowseDiffMergeTool(string toolName, string path)
         {
-            if (IsLoadingSettings)
+            var diffMergeToolConfig = _diffMergeToolConfigurationManager.LoadDiffMergeToolConfig(toolName, null);
```
If the path does not exist (like if the tool is not configured in GE), exception occurs

RussKie replied:
Could you please provide me with steps to reproduce?

![image](https://user-images.githubusercontent.com/4403806/69527155-ccf50000-0fbf-11ea-9c8e-6f23f705fe1e.png)

RussKie replied:
Found the issue, thank you!

With a fix will open the `Program Files` folder, if any issues:
![image](https://user-images.githubusercontent.com/4403806/69527856-3f1a1480-0fc1-11ea-907f-866aa603e017.png)

gerhardol commented on GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.cs:
```diff
@@ -133,118 +137,87 @@ protected override void PageToSettings()
+            if (sender == DifftoolPath)
             {
-                exeFile = MergeToolsHelper.FindDiffToolFullPath(ConfigFileSettingsSet, _NO_TRANSLATE_GlobalDiffTool.Text, out exeName);
+                DifftoolPath.Text = DifftoolPath.Text.ToPosixPath();
```
If you type a \, the cursor is moved back to first position

RussKie replied:
Sorry, I am unable to repro, perhaps I'm not doing it like you are
Please provide more info

gerhardol replied:
Type a path, when entering a '\' the cursor moves to first position
Why not use LostFocus instead on OnChange?

RussKie replied:
Fixed, thank you

gerhardol commented on GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.cs:
```diff
@@ -133,118 +137,87 @@ protected override void PageToSettings()
-            MergeToolCmd.Text = CurrentSettings.GetValue($"mergetool.{mergeTool}.cmd");
+            var filter = !string.IsNullOrWhiteSpace(diffMergeToolConfig.ExeFileName)
+                ? string.Format("{0}|{1}", toolName, diffMergeToolConfig.ExeFileName)
+                : string.Format("{0} (*.exe;*.cmd;*.bat)|*.exe;*.cmd;*.bat", toolName);
```
As the toolname is set to 'toolname.exe' for unknown tools, it can be hard to find the exe
Should probably be changed in 
        public DiffMergeToolConfiguration LoadDiffMergeToolConfig(string toolName, string userSuppliedPath)

RussKie replied:
Could you please clarify?

gerhardol replied:
If I add my tool specificTool, the filter is set to "specificTool.exe" which may not be what I expect

RussKie replied:
Gotcha.

RussKie replied:
fixed in 68ce7dcfbf88c762ef0ff47d63be42be59241bd0

RussKie commented:
> * Remove hardcoded use of .cmd and .path in FormResolveConflicts, the parameters are not required. A few manipulations for instance for Use2WayMerge() will have to be dropped. Or only provide those options when .cmd is available.

Thank you, I totally missed these.

I pushed https://github.com/gitextensions/gitextensions/pull/7044/commits/17f6eec67db54a3874becbc93e2a824e4cbb400b as a prototype.

RussKie commented on GitCommands/DiffMergeTools/Araxis.cs:
```diff
@@ -1,24 +0,0 @@
-using System.Collections.Generic;
```
Chose to remove Araxis as it isn't fully configured as diff/merge tool. I've looked at the tool's configuration and it didn't look trivial.
We can add the tool back, if users request it.

gerhardol replied:
araxis is supported by Git natively, so .cmd is not needed

gerhardol replied:
https://www.araxis.com/merge/documentation-windows/integrating-with-other-applications.en#GitExtensions

The builtin support in master is not up to date
Remove 6.5 references, released 2003-2006

gerhardol replied:
I still believe Araxis should be kept, see the configuration in the link.
Cannot be used much though...

gerhardol commented:
> Thank you, I totally missed these.
I pushed 17f6eec as a prototype.

I wrote that down primary as an indication of a followup PR, not necessarily here.
Some parts that I planned to write already was implemented though (like editing the arguments), so the list was short.

However, _if_ 2way merge is kept as a feature, manipulation must still be done on mergetool.*.cmd arguments, the user should be allowed to change them (that can be done already, from the command line). The feature will then only be available if .cmd exists.

RussKie commented:
> I wrote that down primary as an indication of a followup PR, not necessarily here.
> Some parts that I planned to write already was implemented though (like editing the arguments), so the list was short.
> 
> However, _if_ 2way merge is kept as a feature, manipulation must still be done on mergetool.*.cmd arguments, the user should be allowed to change them (that can be done already, from the command line). The feature will then only be available if .cmd exists.

Then I'd rather leave it as is for now, if we're ok with that. I will add a comment.

RussKie commented on GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.cs:
```diff
@@ -254,23 +242,53 @@ private void GlobalDiffTool_TextChanged(object sender, EventArgs e)
+            var toolName = _NO_TRANSLATE_GlobalDiffTool.Text;
+
+            DifftoolPath.Enabled =
+                BrowseDiffTool.Enabled =
```
What would it make less confusing?
Personally I prefer a Hungarian notation for UI control names, e.g. `txtDiffToolPath`, `btnBrowseDiffTool` etc, but it isn't prevalent in our codebase.

gerhardol replied:
In this case it is three properties that start with Difftool* and one that start with BrowseDiifftool, so evey time I read this, my eyes get stuck and I look for a problem.
Not requesting a a change.

RussKie commented on GitCommands/DiffMergeTools/TortoiseGitMerge.cs:
```diff
@@ -0,0 +1,23 @@
+        public override IEnumerable<string> SearchPaths => new[]
+        {
+            @"TortoiseGit\bin\",
+            @"TortoiseSVN\bin\"
```
done

RussKie commented on GitCommands/DiffMergeTools/TortoiseGitMerge.cs:
```diff
@@ -0,0 +1,23 @@
+        public override string MergeCommand => "-base:\"$BASE\" -mine:\"$LOCAL\" -theirs:\"$REMOTE\" -merged:\"$MERGED\"";
+
+        /// <inheritdoc />
+        public override string Name => "tmerge";
```
done

RussKie commented on GitCommands/DiffMergeTools/BeyondCompare4.cs:
```diff
@@ -0,0 +1,20 @@
+        public override string ExeFileName => "bcomp.exe";
+
+        /// <inheritdoc />
+        public override string Name => "beyondcompare4";
```
1. We have beyondcompare3 and beyondcompare4
2. This is what users see in the dropdowns. I don't think "bc" is very descriptive

RussKie commented on GitCommands/DiffMergeTools/BeyondCompare3.cs:
```diff
@@ -0,0 +1,20 @@
+        public override string ExeFileName => "bcomp.exe";
+
+        /// <inheritdoc />
+        public override string Name => "beyondcompare3";
```
Ok

RussKie commented on GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.cs:
```diff
@@ -254,23 +242,53 @@ private void GlobalDiffTool_TextChanged(object sender, EventArgs e)
+            {
+                DifftoolPath.Text =
+                    DifftoolCmd.Text = string.Empty;
+                return;
```
Thank you, though the missing flow is in `PageToSettings` - if either `difftool` or `mergetool` aren't set we need to zero out respective configurations.

RussKie replied:
done

RussKie commented:
I believe I addressed all comments. Ready for more :)

gerhardol commented on GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.cs:
```diff
@@ -133,118 +144,124 @@ protected override void PageToSettings()
-                exeFile = MergeToolsHelper.FindMergeToolFullPath(ConfigFileSettingsSet, _NO_TRANSLATE_GlobalMergeTool.Text, out exeName);
+                try
+                {
+                    string initialDirectory = Path.GetDirectoryName(path);
```
If path is a directory without a slash, the directory will be cleaned. Suggest the following:

                    if (Directory.Exists(path))
                    {
                        return PathUtil.ToNativePath(PathUtil.EnsureTrailingPathSeparator(path));
                    }

RussKie commented:
> I still believe Araxis should be kept, use the configuration in the link

I don't want to restore it because it was not fully configured.
Any user who has it configured should see no changes, however if we want to add Araxis as a first class citizen it needs to be configured. This is a separate piece of work.

gerhardol commented on GitCommands/DiffMergeTools/VsDiffMerge.cs:
```diff
@@ -32,11 +32,13 @@ private static string GetVsDiffMergePath()
                 return ExeName;
             }
 
+            // For 2017 (15.0) and later, Common7 is not installed by default but often included
```
For 2017 (15.0) and later, VsDiffMerge is not installed by default but often included

RussKie replied:
fixed

gerhardol commented on GitCommands/DiffMergeTools/DiffMergeToolConfigurationManager.cs:
```diff
@@ -0,0 +1,237 @@
+            }
+            else
+            {
+                var pathsToSearch = new[] { UnquoteString(GetToolSetting(diffTool.Name, DiffMergeToolType.Merge, "path")) }.Union(diffTool.SearchPaths);
```
This will get the path to the mergetool with the suggested name, which may be different from the difftool
Normally this will work (as mergetool is searched also for difftool)

RussKie replied:
What are the changes of difftool executable be different from a mergetool executable?
From my observations all tools I've seen use the same executable.

gerhardol commented on GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.cs:
```diff
@@ -133,118 +144,125 @@ protected override void PageToSettings()
-            MergeToolCmd.Text = CurrentSettings.GetValue($"mergetool.{mergeTool}.cmd");
-
-            MergeToolCmd.Enabled = !string.IsNullOrEmpty(MergeToolCmd.Text);
+            string initialDirectory = GetInitialDirectory(path.EnsureTrailingPathSeparator(),
```
Both path and diffMergeToolConfig.Path may be the full path to the executable, this method does not check that the path is a directory

RussKie replied:
I'm due to write tests for this functionality

RussKie replied:
Done. The code is updated to cater for both directory and file

RussKie commented:
> * A path like c:\ is not accepted as a directory

Opps, midnight coding sucks... I'll fix that.

> * If .path is not set, cmd should not be set either. Can be seen as a user error though...

Thank you! I'll fix it

gerhardol commented:
> Opps, midnight coding sucks... I'll fix that.

Midnight reviews are even worse - this PR has plenty of examples

RussKie commented:
> * A path like c:\ is not accepted as a directory
> * If .path is not set, cmd should not be set either. Can be seen as a user error though...

Updated

gerhardol commented:
Still some issues:
Paths ending with / will not work in the file dialog, as dialog only accepts Windows file separators.
CalculateInitialDirectory() should transform to Windows native paths with ToNativePath() (maybe an issue in OpenFileDialog....)
(OpenFileDialog uses a recent path that seem almost random to me, so this appears as weird behaiour for the user.)

If you press Browse, the Cancel it takes some seconds to get back control. The delay is in OpenFileDialog, not sure if something can/should be done about it.

A partial path is added to the cmd, no evaluation if the path is a file, this is fine.
However, an empty path will be be propagated with Suggest.
I could accept that as a feature (to see the arguments).

RussKie commented:
> However, an empty path will be be propagated with Suggest.
> I could accept that as a feature (to see the arguments).

This is intentional, precisely for the reason - to see the arguments.

> CalculateInitialDirectory() should transform to Windows native paths with ToNativePath() (maybe an issue in OpenFileDialog....)

It works for posix-formatted paths. I tried:
* C:/Program Files/Perforce
![image](https://user-images.githubusercontent.com/4403806/70425338-26cbee80-1ac5-11ea-9539-0760297cfabc.png)
* C:/Program Files/Perforce/
![image](https://user-images.githubusercontent.com/4403806/70425220-f71ce680-1ac4-11ea-89f5-d4cfb78ac7ac.png)
* C:/Program Files/Perforce/p4merge.exe
![image](https://user-images.githubusercontent.com/4403806/70425397-3e0adc00-1ac5-11ea-88ff-ee49ab6a3e3f.png)

Is there some other use-case I'm missing?


> OpenFileDialog uses a recent path that seem almost random to me, so this appears as weird behaiour for the user.

The initial directory is calculated as follows:
* user-provided path to a folder, if it exists, or
* the selected tool preferred path (the first existing, if there are alternatives), or
* ProgramFiles folder

gerhardol commented:
With dir set to c:/, I get "C:/Program Files/Perforce" (mergetool path, the last selected path).
The debugger did show initial path as "c:/" as expected, same for "c:\" also as expected. "C:" is converted to "c:\" and is OK.

So this is a specific case that can be disregarded, this is an issue with OpenFileDialog.
